### PR TITLE
fix calling id.watch() from the head

### LINKED ISSUE #2252: Calling navigator.id.watch() from the document <head> breaks onlogin/onlogout
If you take the updated [Python cookbook example](https://github.com/mozilla/browserid-cookbook/blob/master/python/python.cgi) and apply the following patch (i.e. moving the `navigator.id.watch()` call from the `body` to the `head` of the page):

```
diff --git a/python/python.cgi b/python/python.cgi
index dba2f07..a559c32 100755
--- a/python/python.cgi
+++ b/python/python.cgi
@@ -10,8 +10,6 @@ def print_header(email = 'null'):

     print """<!DOCTYPE html><html><head><meta charset="utf-8">
 <script src="https://login.persona.org/include.js"></script>
-</head>
-<body>
 <script>
 navigator.id.watch({
     loggedInEmail: %s,
@@ -26,6 +24,8 @@ navigator.id.watch({
     },
 });
 </script>
+</head>
+<body>
 <form id="login-form" method="POST">
 <input id="assertion-field" type="hidden" name="assertion" value="">
 </form>""" % email
```

This problem occurs because:
1. `navigator.id.logout()` won't do anything if [`loggedInUser` isn't set properly](https://github.com/mozilla/browserid/blob/26a7264a31be3c210d7257bd0ab785ba5c040d3f/resources/static/communication_iframe/start.js#L96)
2. `loggedInUser` isn't set in the `navigator.id.watch()` call if [the communication channel isn't open](https://github.com/mozilla/browserid/blob/26a7264a31be3c210d7257bd0ab785ba5c040d3f/resources/static/include_js/include.js#L1071)
3. the communication channel is not open because while we're in the `head` of the document, [`dom.body` is null](https://github.com/mozilla/browserid/blob/26a7264a31be3c210d7257bd0ab785ba5c040d3f/resources/static/include_js/include.js#L998)

### PR DESCRIPTION
**Pull request by seanmonstar**
stores loggedInUser into a local variable, and notifies
the Channel whenever it finally builds

fixes #2252

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
@seanmonstar - I ran the PR against a modified example RP, see the attached diff.  @fmarier's observation that document.body is null is still a problem, this prevents the communication iframe from being created if navigator.id.watch is called within the head of the document.

This is not a problem in the initial RP's tab because the assertion is generated in the dialog and passed back through WinChan, making everything appear to operate correctly in the initial tab. There _is_ a problem if the user creates a new account or adds an email to an existing account and they must verify their email. When the email verification page is redirected back to the RP, navigator.id.watch tries to create the communication iframe to generate the assertion. Since document.body does not yet exist, the iframe is not created and no assertion is ever generated.

Perhaps doing a poll for the existence of document.body? Waiting for DOMContentLoaded would work in browsers that support it, for IE8 we would have to use onload. The danger of waiting for onload is that it waits for all document resources to finish, if for some reason an image takes a long time to download or times out, the user could be waiting for an extremely long time.

diff:
http://pastebin.mozilla.org/1768337

**Comment by seanmonstar** [comment]
Oh I see what you mean now. I had thought similarly, but when I tested, I was looking at the console output. The callbacks all being triggered, since the observers are held onto regardless of the iframe. And I was clicking logout first, which tries to add the iframe again.

I'll listen for `DOMContentLoaded` and `readyStateChange` for IE8.

File: resources/static/include_js/include.js [hunk 1]
```diff
@@ -981,18 +983,48 @@
...
+        document.addEventListener('DOMContentLoaded', function contentLoaded() {
+          document.removeEventListener('DOMContentLoaded', contentLoaded);
+          callback();
+        }, false);
+      } else if (document.attachEvent && document.readyState) {
+        document.attachEvent('onreadystatechange', function ready() {
+          var state = document.readyState;
+          // 'interactive' is the same as DOMContentLoaded,
+          // but not all browsers use it, sadly.
+          if (['loaded', 'complete', 'interactive'].indexOf(state) > -1) {
```
Comment (shane-tomlinson) [inline resources/static/include_js/include.js]: IE8 unfortunately doesn't support Array.prototype.indexOf.

Comment (shane-tomlinson) [inline resources/static/include_js/include.js]: @seanmonstar - if I apply the following diff, IE8 acts as expected:

```
diff --git a/resources/static/include_js/include.js b/resources/static/include_js/include.js
index 979bca2..013f05f 100644
--- a/resources/static/include_js/include.js
+++ b/resources/static/include_js/include.js
@@ -997,9 +997,12 @@
           var state = document.readyState;
           // 'interactive' is the same as DOMContentLoaded,
           // but not all browsers use it, sadly.
-          if (['loaded', 'complete', 'interactive'].indexOf(state) > -1) {
-            document.detachEvent('onreadystatechange', ready);
-            callback();
+          var events = ['loaded', 'complete', 'interactive'];
+          for(var i = 0, eventName; eventName = events[i]; ++i) {
+            if(eventName === state) {
+              document.detachEvent('onreadystatechange', ready);
+              callback();
+            }
           }
         });
       }
```

Comment (seanmonstar) [inline resources/static/include_js/include.js]: I forget that so often...

I'm inclined to switch it to `if (state === 'loaded' || state === 'complete' || state === 'interactive')`

**Comment by seanmonstar** [comment]
@shane-tomlinson changed away from indexOf

File: resources/static/include_js/include.js [hunk 2]
```diff
@@ -981,18 +983,48 @@
...
+          document.removeEventListener('DOMContentLoaded', contentLoaded);
+          callback();
+        }, false);
+      } else if (document.attachEvent && document.readyState) {
+        document.attachEvent('onreadystatechange', function ready() {
+          var state = document.readyState;
+          // 'interactive' is the same as DOMContentLoaded,
+          // but not all browsers use it, sadly.
+          if (state === 'loaded' || state === 'complete' || state === 'interactive') {
+            document.detachEvent('onreadystatechange', ready);'
```
Comment (shane-tomlinson) [inline resources/static/include_js/include.js:1001]: There is an extra ' on the end of this line.

File: resources/static/include_js/include.js [hunk 3]
```diff
@@ -981,18 +983,48 @@
...
+        document.addEventListener('DOMContentLoaded', function contentLoaded() {
+          document.removeEventListener('DOMContentLoaded', contentLoaded);
+          callback();
+        }, false);
+      } else if (document.attachEvent && document.readyState) {
+        document.attachEvent('onreadystatechange', function ready() {
+          var state = document.readyState;
+          // 'interactive' is the same as DOMContentLoaded,
+          // but not all browsers use it, sadly.
+          if (state === 'loaded' || state === 'complete' || state === 'interactive') {
```
Comment (shane-tomlinson) [inline resources/static/include_js/include.js:1000]: Much simpler than my solution.

**Comment by shane-tomlinson** [comment]
@seanmonstar - I removed the extra ' and manually merged this in with commit 18c7a2f7c486a9f722fedf1f8d40b9cdef589cf5.  

close #2252

**Pull request merged**